File: 백준/Gold/1103. 게임/게임.cc
#include <bits/stdc++.h>
using namespace std;

const int dy[] = {-1, 0, 1, 0};
const int dx[] = {0, 1, 0, -1};
int n, m;
int visited[54][54], dp[54][54];
char a[54][54];

int go(int y, int x) {
  if (y < 0 || x < 0 || y >= n || x >= m || a[y][x] == 'H') return 0;
  if (visited[y][x]) {
    cout << -1 << '\n';
    exit(0);
  }

  int& ret = dp[y][x];
  if (ret) return ret;

  visited[y][x] = 1;

  for (int i = 0; i < 4; i++) {
    int ny = y + dy[i] * (a[y][x] - '0');
    int nx = x + dx[i] * (a[y][x] - '0');

    ret = max(ret, go(ny, nx) + 1);
  }

  visited[y][x] = 0;
  return ret;
}

int main() {
  cin >> n >> m;
  for (int i = 0; i < n; i++) {
    string temp;
    cin >> temp;
    for (int j = 0; j < m; j++) {
      a[i][j] = temp[j];
    }
  }

  cout << go(0, 0) << '\n';

  return 0;
}

File: 백준/Gold/1103. 게임/게임.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : './input.txt').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const a = input.slice(1).map(line => line.split(''));

const dir = [[-1, 0], [0, 1], [1, 0], [0, -1]];
const visited = Array.from({length: n}, () => Array(m).fill(0));
const dp = Array.from({length: n}, () => Array(m).fill(-1));

function go (y, x) {
  if(y < 0 || x < 0 || y >= n || x >= m || a[y][x] === 'H') return 0;
  if(visited[y][x]){
    console.log(-1);
    process.exit(0);
  }

  if(dp[y][x] !== -1) return dp[y][x];
  let ret = 0;

  visited[y][x] = 1;
  for (const [dy, dx] of dir)  {
    let ny = y + dy * a[y][x], nx = x + dx * a[y][x];

    ret = Math.max(go(ny, nx) + 1, dp[y][x]);
    dp[y][x] = ret;
  }
  visited[y][x] = 0;

  return ret;
}

function answer() {
  console.log(go(0, 0));
}

answer();